openhands/server/routes/microagent.py

```diff
@@ -0,0 +1,215 @@
+2. Provide context for why these updates are necessary based on the new events.
+3. Be specific about the information that should be added or modified.
+4. Maintain the existing structure and formatting of the file.
+5. Ensure that the updates are consistent with the current content and don't contradict existing information.
+
+Now, based on the new events provided, generate a prompt that will guide the AI in making the appropriate updates to the special file. Your prompt should be clear, specific, and actionable. Include your prompt within <update_prompt> tags.
+
+<update_prompt>
+
+</update_prompt>
+"""
+
+
+def get_context_events(
+        events: list[Event],
+        event_id: int,
+        context_size: int = 4,
+) -> list[Event]:
+    """
+    Get a list of events around a specific event ID.
+
+    Args:
+        events: List of events to search through.
+        event_id: The ID of the target event.
+        context_size: Number of events to include before and after the target event.
+
+    Returns:
+        A list of events including the target event and the specified number of events before and after it.
+    """
+    target_event_index = None
+    for i, event in enumerate(events):
+        if event.id == event_id:
+            target_event_index = i
+            break
+        
+    if target_event_index is None:
+        raise ValueError(f"Event with ID {event_id} not found in the event stream.")
+
+    # Get X events around the target event
+    start_index = max(0, target_event_index - context_size)
+    end_index = min(len(events), target_event_index + context_size + 1) # +1 to include the target event
+
+    return events[start_index:end_index]
+
+
+def generate_prompt(llm_config: LLMConfig, events: str, prompt_template: str = prompt_template) -> str:
+    llm = LLM(llm_config)
+    messages = [
+        {
+            "role": "system",
+            "content": prompt_template.replace("{{EVENTS}}", events),
+        },
+        {
+            "role": "user",
+            "content": "Please generate a prompt for the AI to update the special file based on the events provided.",
+        },
+    ]
+
+    response = llm.completion(messages=messages)
+    raw_prompt = response["choices"][0]["message"]["content"].strip()
+    prompt = re.search(
+        r"<update_prompt>(.*?)</update_prompt>", raw_prompt, re.DOTALL
+    )
+
+    if prompt:
+        return prompt.group(1).strip()
+    else:
+        raise ValueError("No valid prompt found in the response.")
+
+
+async def generate_unique_conversation_id(
+        conversation_store: ConversationStore,
+) -> str:
+    conversation_id = uuid.uuid4().hex
+    while await conversation_store.exists(conversation_id):
+        conversation_id = uuid.uuid4().hex
+    return conversation_id
+
+app = APIRouter(prefix='/knowledge')
+
+class KnowledgePromptRequest(BaseModel):
+    conversation_id: str
+    event_id: int
+
+
+@app.get('/prompt/{conversation_id}')
+async def get_prompt(
```

Comment:
I'm not sure this will kick our conversation middleware into gear, meaning it might by pass auth. We should generally use the EvenStream that gets attached to the request object rather than constructing it dynamically

I think if you put this on `/{conversation_id/remember_prompt` it'll work better. Than you can use `request.state.conversation.event_stream`